core/src/layers/blocking.rs

```diff
@@ -0,0 +1,262 @@
+        self.runtime.block_on(self.inner.create_dir(path, args))
+    }
+
+    fn blocking_read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::BlockingReader)> {
+        self.runtime.block_on(async {
+            let (rp, reader) = self.inner.read(path, args).await?;
+            let blocking_reader = BlockingReader::new(self.runtime.clone(), reader);
+
+            Ok((rp, blocking_reader))
+        })
+    }
+
+    fn blocking_write(&self, path: &str, args: OpWrite) -> Result<(RpWrite, Self::BlockingWriter)> {
+        self.runtime.block_on(async {
+            let (rp, writer) = self.inner.write(path, args).await?;
+            let blocking_writer = Self::BlockingWriter::new(self.runtime.clone(), writer);
+            Ok((rp, blocking_writer))
+        })
+    }
+
+    fn blocking_copy(&self, from: &str, to: &str, args: OpCopy) -> Result<RpCopy> {
+        self.runtime.block_on(self.inner.copy(from, to, args))
+    }
+
+    fn blocking_rename(&self, from: &str, to: &str, args: OpRename) -> Result<RpRename> {
+        self.runtime.block_on(self.inner.rename(from, to, args))
+    }
+
+    fn blocking_stat(&self, path: &str, args: OpStat) -> Result<RpStat> {
+        self.runtime.block_on(self.inner.stat(path, args))
+    }
+
+    fn blocking_delete(&self, path: &str, args: OpDelete) -> Result<RpDelete> {
+        self.runtime.block_on(self.inner.delete(path, args))
+    }
+
+    fn blocking_list(&self, path: &str, args: OpList) -> Result<(RpList, Self::BlockingPager)> {
+        self.runtime.block_on(async {
+            let (rp, pager) = self.inner.list(path, args).await?;
+            let blocking_pager = Self::BlockingPager::new(self.runtime.clone(), pager);
+            Ok((rp, blocking_pager))
+        })
+    }
+}
+
+pub struct BlockingReader<R> {
+    runtime: Arc<runtime::Runtime>,
+    inner: R,
+}
+
+impl<R> BlockingReader<R> {
+    fn new(runtime: Arc<runtime::Runtime>, inner: R) -> Self {
+        Self { runtime, inner }
+    }
+}
+
+// impl<R: oio::Read> oio::Read for BlockingReader<R> {
+//     fn poll_read(&mut self, cx: &mut Context<'_>, buf: &mut [u8]) -> Poll<Result<usize>> {
+//         self.inner.poll_read(cx, buf)
+//     }
+
+//     fn poll_seek(&mut self, cx: &mut Context<'_>, pos: io::SeekFrom) -> Poll<Result<u64>> {
+//         self.inner.poll_seek(cx, pos)
+//     }
+
+//     fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
+//         self.inner.poll_next(cx)
+//     }
+// }
+
+impl<R: oio::Read + 'static> oio::BlockingRead for BlockingReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        self.runtime
+            .block_on(oio::ReadExt::read(&mut self.inner, buf))
+    }
+
+    fn seek(&mut self, pos: std::io::SeekFrom) -> Result<u64> {
+        self.runtime.block_on(self.inner.seek(pos))
+    }
+
+    fn next(&mut self) -> Option<Result<Bytes>> {
+        self.runtime.block_on(self.inner.next())
+    }
+}
+
+pub struct BlockingWriter<W> {
```

Comment:
We can implement a `BlockingWrapper` and implement different traits for it.

Reply:
Great idea!